process.go

```diff
@@ -12,40 +12,43 @@ import (
 
 // LaunchProcess runs a subprocess and returns when the subprocess exits,
 // either when it dies, or *after* a successful upgrade.
-func LaunchProcess(cfg *Config, args []string, stdout, stderr io.Writer) error {
+func LaunchProcess(cfg *Config, args []string, stdout, stderr io.Writer) (bool, error) {
+	var doUpgrade bool
 	bin := cfg.CurrentBin()
+
 	err := EnsureBinary(bin)
 	if err != nil {
-		return errors.Wrap(err, "current binary invalid")
+		return doUpgrade, errors.Wrap(err, "current binary invalid")
 	}
 
 	cmd := exec.Command(bin, args...)
 	outpipe, err := cmd.StdoutPipe()
 	if err != nil {
-		return err
+		return doUpgrade, err
 	}
 	errpipe, err := cmd.StderrPipe()
 	if err != nil {
-		return err
+		return doUpgrade, err
 	}
 	scanOut := bufio.NewScanner(io.TeeReader(outpipe, stdout))
 	scanErr := bufio.NewScanner(io.TeeReader(errpipe, stderr))
 
 	err = cmd.Start()
 	if err != nil {
-		return errors.Wrapf(err, "launching process %s %s", bin, strings.Join(args, " "))
+		return doUpgrade, errors.Wrapf(err, "launching process %s %s", bin, strings.Join(args, " "))
 	}
 
 	// three ways to exit - command ends, find regexp in scanOut, find regexp in scanErr
 	upgradeInfo, err := WaitForUpgradeOrExit(cmd, scanOut, scanErr)
 	if err != nil {
-		return err
+		return doUpgrade, err
 	}
 	if upgradeInfo != nil {
-		return DoUpgrade(cfg, upgradeInfo)
+		doUpgrade = true
+		return doUpgrade, DoUpgrade(cfg, upgradeInfo)
```

Comment:
You can simplify this to `return true, DoUpgrade(cfg, upgradeInfo)`

and let the final case remain as `return false, nil`

Reply:
Yes, that will be straight forward. I will update